src/smartpeak/source/io/SessionDB.cpp

```diff
@@ -193,25 +193,292 @@ namespace SmartPeak
+{
+  auto table_name = properties_handler.getPropertiesHandlerName();
+  LOGD << "Writting " << table_name << " to session db.";
+
+  std::ostringstream os;
+  int rc;
+  char* zErrMsg = nullptr;
+  DBContext db_context;
+
+  // open DB
+  auto db = openSessionDB();
+  if (!db)
+  {
+    return false;
+  }
+
+  updateSessionInfo(*db);
+
+  db_context.table = table_name;
+  db_context.db = *db;
+
+  // first, we delete table (to remove)
+  os.str("");
+  os << "DROP TABLE ";
+  os << "\"" << table_name << "\"";
+  os << ";";
+  std::string sql = os.str();
+  rc = sqlite3_exec(*db, sql.c_str(), NULL, 0, &zErrMsg);
+  // -- ignore error
+
+  // create table
+  auto properties = properties_handler.getPropertiesSchema();
+  os.str("");
+  os << "CREATE TABLE ";
+  os << "\"" << table_name << "\"";
+  os << " (";
+  os << "ID INTEGER PRIMARY KEY, ";
+  std::string separator;
+  for (const auto& [column_name, column_type] : properties)
+  {
+    // beginWrite(os, db_context.columns, value, value_type, args...);
+    os << separator;
+    os << "\"" << column_name << "\"";
+    os << " ";
+    os << valueTypeToString(column_type);
+    os << "  NOT NULL";
+    separator = ", ";
+  }
+  os << "); ";
+  sql = os.str();
+  rc = sqlite3_exec(*db, sql.c_str(), NULL, 0, &zErrMsg);
+  if (rc != SQLITE_OK)
+  {
+    logSQLError(zErrMsg, sql);
+    sqlite3_free(zErrMsg);
+    return false;
   }
 
-  void SessionDB::logSQLError(const std::string& error_message, const std::string& sql_command) const
+  // write data
+  os.str("");
+  auto nb_rows = properties_handler.getNbRows();
+  for (size_t i = 0; i < nb_rows; ++i)
   {
-    LOGE << "SQL Error: " << error_message;
-    if (!sql_command.empty())
+    os << "INSERT INTO ";
+    os << "\"" << table_name << "\"";
+    separator = "";
+    os << " (";
+    for (const auto& [column_name, column_type] : properties)
     {
-      LOGE << "SQL Command: " << sql_command;
+      os << separator;
+      os << "\"" << column_name << "\"";
+      separator = ", ";
     }
+    os << ") ";
+
+    os << "VALUES (";
+    separator = "";
+    for (const auto& [column_name, column_type] : properties)
+    {
+      os << separator;
+      auto value = properties_handler.getProperty(column_name);
+      if (value)
+      {
+        os << castValueToSqlString(*value);
+      }
+      separator = ", ";
+    }
+    os << "); ";
+    std::string sql = os.str();
+    rc = sqlite3_exec(db_context.db, sql.c_str(), NULL, 0, &zErrMsg);
+    if (rc != SQLITE_OK) {
+      logSQLError(zErrMsg, sql);
+      sqlite3_free(zErrMsg);
+      return false;
+    }
+  }
+
+  // end write
+  if (db_context.stmt)
+  {
+    sqlite3_finalize(db_context.stmt);
+    db_context.stmt = nullptr;
   }
+  closeSessionDB(db_context.db);
 
-  int64_t SessionDB::getLastInsertedRowId(SessionDB::DBContext& db_context) const
+  return true;
+}
+
+bool SessionDB::readPropertiesHandler(IPropertiesHandler& properties_handler)
```

Comment:
The read/writePropertiesHandler introduces a lot of new code to the class.  Are you sure it is not appropriate to reuse the code you had already written for reading and writing the the SQLite DB?

Reply:
the other set of API is actually not so easy to reuse as it is templated, unfortunately. I would suggest one day unifying this with one set of API (the new one is better)